vortex-file/src/segments/driver.rs

```diff
@@ -0,0 +1,413 @@
+                        .request
+                        .take()
+                        .vortex_expect("checked above as present"),
+                );
+            }
+
+            // If we added no new segments, we're done.
+            if coalesced.requests.len() == request_count {
+                break;
+            }
+        }
+
+        // Ensure the coalesced requests are sorted
+        coalesced.requests.sort_by_key(|r| r.id());
+
+        // Maintain the prefetch buffer count.
+        for request in &coalesced.requests {
+            self.mark_as_prefetched(request.id());
+        }
+
+        log::debug!("Coalesced request: {:?}", coalesced);
+        self.coalesced_counter.inc();
+        self.coalesced_bytes_counter
+            .add(coalesced.size_bytes().try_into().vortex_expect("isize"));
+        coalesced
+    }
+}
+
+impl Stream for CoalescedDriver {
+    type Item = CoalescedSegmentRequest;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        // First, we consume any events that have been emitted.
+        while let Poll::Ready(event) = this.events.poll_next_unpin(cx) {
+            let Some(event) = event else {
+                // The event source has shut down, we can end the stream.
+                return Poll::Ready(None);
+            };
+
+            // Process the event.
+            log::debug!("Processing: {:?}", event);
+            match event {
+                SegmentEvent::Requested(req) => this.on_requested(req),
+                SegmentEvent::Polled(id) => this.on_polled(id),
+                SegmentEvent::Dropped(id) => this.on_dropped(id),
+                SegmentEvent::Resolved(id) => this.on_resolved(id),
+            }
+        }
+
+        // If no segments have been polled yet, then we hold off on launching I/O.
+        if !this.first_poll {
+            return Poll::Pending;
+        }
+
+        // Now we check to see if we should launch any I/O.
+        while let Some(id) = this.polled.pop_front() {
+            if let Some(request) = this
+                .state
+                .get_mut(&id)
+                .and_then(|state| state.request.take())
+            {
+                return Poll::Ready(Some(this.coalesce_request(request)));
+            }
+        }
+
+        // Only perform pre-fetching if we have spare capacity.
+        log::trace!(
+            "Used {} / {} prefetched bytes",
+            this.prefetched_bytes,
+            this.max_prefetch_bytes
+        );
+        if this.prefetched_bytes < this.max_prefetch_bytes {
+            while let Some(id) = this.requested.pop_front() {
+                if let Some(request) = this
+                    .state
+                    .get_mut(&id)
+                    .and_then(|state| state.request.take())
+                {
+                    let coalesced = this.coalesce_request(request);
```

Comment:
Coalescing currently doesn't care about polled vs requested segments. Maybe we can revisit this later and prioritise coalescing towards polled segments. If on the left side of the current segment there are more polled segments instead of widening the request on both sides we can bias towards the left for example